extensions/expression-language/Expression.cpp

```diff
@@ -0,0 +1,164 @@
+  std::transform(result.begin(), result.end(), result.begin(), ::toupper);
+  return result;
+}
+
+template<std::string T(const std::vector<std::string> &)>
+Expression make_dynamic_function_incomplete(const std::string &function_name,
+                                            const std::vector<Expression> &args,
+                                            std::size_t num_args) {
+  if (args.size() == num_args) {
+    return make_dynamic([=](const Parameters &params) -> std::string {
+      std::vector<std::string> evaluated_args;
+
+      for (const auto &arg : args) {
+        evaluated_args.emplace_back(arg(params));
+      }
+
+      return T(evaluated_args);
+    });
+  } else {
+    auto result = make_dynamic([](const Parameters &params) -> std::string {
+      throw std::runtime_error("Attempted to call incomplete function");
+    });
+
+    result.complete = [function_name, args](Expression expr) -> Expression {
+      std::vector<Expression> complete_args = {expr};
+      complete_args.insert(complete_args.end(), args.begin(), args.end());
+      return make_dynamic_function(function_name, complete_args);
+    };
+
+    return result;
+  }
+}
+
+Expression make_dynamic_function(const std::string &function_name, const std::vector<Expression> &args) {
+  if (function_name == "hostname") {
+    return make_dynamic_function_incomplete<expr_hostname>(function_name, args, 0);
+  } else if (function_name == "toUpper") {
+    return make_dynamic_function_incomplete<expr_toUpper>(function_name, args, 1);
+  } else {
+    std::string msg("Unknown expression function: ");
+    msg.append(function_name);
+    throw std::runtime_error(msg);
+  }
+}
+
+Expression make_dynamic_function_postfix(const Expression &subject, const Expression &fn) {
+  return fn.complete(subject);
+}
+
+bool Expression::isDynamic() const {
+  if (val_fn_) {
+    return true;
+  } else {
+    return false;
+  }
+}
+
+Expression Expression::operator+(const Expression &other_expr) const {
+  if (isDynamic() && other_expr.isDynamic()) {
+    auto val_fn = val_fn_;
+    auto other_val_fn = other_expr.val_fn_;
+    return make_dynamic([val_fn, other_val_fn](const Parameters &params) -> std::string {
+      std::string result = val_fn(params);
+      result.append(other_val_fn(params));
+      return std::move(result);
+    });
+  } else if (isDynamic() && !other_expr.isDynamic()) {
+    auto val_fn = val_fn_;
+    auto other_val = other_expr.val_;
+    return make_dynamic([val_fn, other_val](const Parameters &params) -> std::string {
+      std::string result = val_fn(params);
+      result.append(other_val);
+      return std::move(result);
+    });
+  } else if (!isDynamic() && other_expr.isDynamic()) {
+    auto val = val_;
+    auto other_val_fn = other_expr.val_fn_;
+    return make_dynamic([val, other_val_fn](const Parameters &params) -> std::string {
+      std::string result(val);
+      result.append(other_val_fn(params));
+      return std::move(result);
```

Comment:
The move prevents copy elision through RVO.

Reply:
"
In a return statement or a throw-expression, if the compiler cannot perform copy elision but the conditions for copy elision are met or would be met, except that the source is a function parameter, the compiler will attempt to use the move constructor even if the object is designated by an lvalue; see return statement for details."